libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -128,238 +169,772 @@ BeaconSetupService::State BeaconSetupService::OnIdle()
+ * @param from_id Muddle address of sender
+ */
+void BeaconSetupService::OnReconstructionShares(SharesMessage const &shares_msg,
+                                                MuddleAddress const &from)
+{
+  // Return if the sender is in complaints, or not in QUAL
+  // TODO(JMW): Could be problematic if qual has not been built yet
+  if (qual_complaints_manager_.ComplaintsFind(from) ||
+      beacon_->manager.qual().find(from) == beacon_->manager.qual().end())
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Node ", beacon_->manager.cabinet_index(),
+                   " received message from invalid sender. Discarding.");
+    return;
+  }
+  if (reconstruction_shares_received_.find(from) == reconstruction_shares_received_.end())
+  {
+    for (auto const &share : shares_msg.shares())
+    {
+      beacon_->manager.VerifyReconstructionShare(from, share);
+    }
+    reconstruction_shares_received_.insert(from);
+  }
+  else
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Node ", beacon_->manager.cabinet_index(),
+                   " received duplicate reconstruction shares from node ",
+                   beacon_->manager.cabinet_index(from));
+  }
+}
+
+/**
+ * For all complaint answers received in defense of a complaint we check the exposed secret share
+ * is consistent with the broadcasted coefficients
+ *
+ * @param answer Pointer of message containing some secret shares of sender
+ * @param from_id Muddle address of sender
+ * @param from_index Index of sender in cabinet_
+ */
+void BeaconSetupService::CheckComplaintAnswer(SharesMessage const &answer,
+                                              MuddleAddress const &from_id)
+{
+  // If not enough answers are sent for number of complaints against a node then add a complaint a
+  // against it
+  auto diff = complaints_manager_.ComplaintsCount(from_id) - answer.shares().size();
+  if (diff > 0 && diff <= beacon_->aeon.members.size())
   {
-    auto &s = t.second;
-    if (!beacon_->manager.AddShare(s.from, s.share, s.verification_vector))
+    complaints_answer_manager_.Add(from_id);
+  }
+  for (auto const &share : answer.shares())
+  {
+    if (!beacon_->manager.VerifyComplaintAnswer(from_id, share))
     {
-      // TODO(tfr): Serializable error
-      throw std::runtime_error("share could not be verified.");
+      complaints_answer_manager_.Add(from_id);
     }
   }
+}
 
-  // and generating secret
-  beacon_->manager.CreateKeyPair();
+/**
+ * Builds the set of qualified members of the cabinet.  Altogether, complaints consists of
+  // 1. Nodes which received over t complaints
+  // 2. Complaint answers which were false
 
-  return State::BEACON_READY;
+ * @return True if self is in qual and qual is at least of size polynomial_degree_ + 1, false
+ otherwise
+ */
+bool BeaconSetupService::BuildQual()
+{
+  // Create set of muddle addresses
+  std::set<MuddleAddress> cabinet;
+  for (auto &m : beacon_->aeon.members)
+  {
+    cabinet.insert(m.identifier());
```

Comment:
if we use set instead of unordered set everywhere this could be an assignment rather than loop, perhaps we should do that

Reply:
Yes, this is the TODO: Fix I had above this. I have pushed this in a later commit.

Reply:
Yes, I agree this would be better but would require RBC to take in a set of identities rather than just muddle addresses. I agree we should do this.